proposals/webidl-bindings/bindings.md

```diff
@@ -0,0 +1,508 @@
+    (access ($Base $Len) (integer-to-i32 
+      (codePointCount (base-len-as-string $Base $Len)))))
+  ...
+)
+```
+</div>
+
+### Export Bindings
+
+The _ExportBindings_ special section defines the bindings for functions that are
+exported by the module. Like imports, exports from a webAssembly module are
+managed on a per function basis.
+
+```
+ExportBinding => 'export-binding' FunctionName ProvisionLambda
+```
+
+The W-Expression form of an export binding is given by:
+
+```
+W-ExportBinding ::= '(' 'export-binding' FunctionName W-ProvisionLambda ')'
+```
+
+<div class="example">
+
+The corresponding module that exports the code point counter alluded to above
+may resemble:
+
+```
+(module
+  ...
+  (export "codePointCount" (func $_cp_count))
+  ...
+  (func $_cp_count (return i32) (param $b i32 $l i32)
+    ...
+  )
+  ...
+  (export-binding "codePointCount"
+    (provide (S) 
+      (i32-as-integer 
+        ($_cp_count ^ (string-to-base-ptr S $buffer $buffLen)))))
+  ...
+)
+```
+
+</div>
+
+## Coercion Operators
+
+The coercion operators are used to lift webAssembly values into wi-IDL values
+and vice versa.
+
+The general form of a coercion operator takes the form of a function call; but as
+previously noted, coercion operators are not conventional functions.
+
+### Lifting Operators
+
+#### i32-as-integer
+
+The `i32-as-integer` operator maps a webAssembly `i32` value into the wi-IDL
+space of `Integer`. 
+
+Note that this does not necessarily imply any _transformation_ of the bits
+representing the integer value.
+
+```
+i32-as-integer : (i32) => Integer
+```
+
+#### base-len-as-string
+
+The `base-len-as-string` operator maps a utf8 encoded sequence of bytes into the wi-IDL space of `String`
+
+```
+base-len-as-string: (i32 i32) => String
+```
+
+### Lowering Operators
+
+Lowering operators are the complement of the lifting operators; generally they
+are used to map values in the wi-IDL space into the webAssembly space.
+
+#### integer-to-i32
+
+The `integer-to-i32` operator maps a wi-IDL `Integer` value into the webAssembly
+space of 32-bit integers.
+
+```
+Integer-to-i32 : (Integer) -> i32
+```
+
+Note that this operator _may_ involve a _truncation_ of values if the original
+source of the `Integer` value had a higher precision than 32 bits.
+
+In particular, the identity
+
+```
+(integer-to-i32 (i32-as-integer X)) == X
+```
+is satisfied, whereas the identity
+
+```
+(integer-to-i32 (i64-as-integer X)) == X
+```
+is not, in general.
+
+#### string-to-base-ptr
+
+The `string-to-base-ptr` maps a wi-IDL `String` value into a block of webAssembly linear memory -- as a sequence of utf-8 encoded bytes.
+
+```
+(string-to-base-ptr : (String i32 i32) => (i32 i32) $throws i32 
+```
+
+The three arguments to `string-to-base-ptr` are the `String` value itself, the
+base address of a block of linear memory -- together with the size of the
+buffer. The returned result is the address of the utf8-encoded sequence of bytes
```

Comment:
Am I understanding correctly that these three arguments are runtime arguments, and none are immediates? Where is the buffer supposed to come from? Are we going to allow some subset of Wasm instructions inside access/provision lambdas so that they can call allocators themselves instead of building the concept of allocators into coercion operators?

If there isn't enough room for the incoming string in the given allocation buffer, what happens? Does this try calling an allocator function? Throw? Only write a substring into the buffer?